src/logic/safe/store/actions/transactions/fetchTransactions/loadGatewayTransactions.ts

```diff
@@ -1,13 +1,96 @@
-import { getTransactionHistory, getTransactionQueue } from '@gnosis.pm/safe-react-gateway-sdk'
+import {
+  getTransactionHistory,
+  getTransactionQueue,
+  TransactionListPage,
+  getIncomingTransfers,
+  getMultisigTransactions,
+  getModuleTransactions,
+} from '@gnosis.pm/safe-react-gateway-sdk'
 import { _getChainId } from 'src/config'
 import { HistoryGatewayResponse, QueuedGatewayResponse } from 'src/logic/safe/store/models/types/gateway.d'
 import { checksumAddress } from 'src/utils/checksumAddress'
 import { Errors, CodedException } from 'src/logic/exceptions/CodedException'
+import { FilterForm, FilterType, FILTER_TYPE_FIELD_NAME } from 'src/routes/safe/components/Transactions/TxList/Filter'
+import {
+  getIncomingFilter,
+  getMultisigFilter,
+  getModuleFilter,
+} from 'src/routes/safe/components/Transactions/TxList/Filter/utils'
+import { ChainId } from 'src/config/chain.d'
+import { operations } from '@gnosis.pm/safe-react-gateway-sdk/dist/types/api'
 
 /*************/
 /*  HISTORY  */
 /*************/
-const historyPointers: { [chainId: string]: { [safeAddress: string]: { next?: string; previous?: string } } } = {}
+const historyPointers: {
+  [chainId: string]: {
+    [safeAddress: string]: {
+      next?: string
+      previous?: string
+      filterType?: FilterType
+    }
+  }
+} = {}
+
+const getHistoryTxListPage = async (
+  chainId: ChainId,
+  safeAddress: string,
+  filter?: FilterForm | Partial<FilterForm>,
+): Promise<TransactionListPage> => {
+  let txListPage: TransactionListPage = {
+    next: undefined,
+    previous: undefined,
+    results: [],
+  }
+
+  const { next, filterType } = historyPointers[chainId]?.[safeAddress] || {}
+
+  let query:
+    | operations['incoming_transfers' | 'incoming_transfers' | 'module_transactions']['parameters']['query']
+    | undefined
+
+  switch (filterType) {
+    case FilterType.INCOMING: {
+      query = filter ? getIncomingFilter(filter) : undefined
+      txListPage = await getIncomingTransfers(chainId, safeAddress, query, next)
+      break
+    }
+    case FilterType.MULTISIG: {
+      query = filter ? getMultisigFilter(filter, true) : undefined
+      txListPage = await getMultisigTransactions(chainId, safeAddress, query, next)
+      break
+    }
+    case FilterType.MODULE: {
+      query = filter ? getModuleFilter(filter) : undefined
+      txListPage = await getModuleTransactions(chainId, safeAddress, query, next)
+      break
+    }
+    default: {
+      txListPage = await getTransactionHistory(chainId, safeAddress, next)
+    }
+  }
+
+  const getPageUrl = (pageUrl?: string): string | undefined => {
+    if (!pageUrl || !filterType || !query) {
+      return pageUrl
+    }
+
+    const url = new URL(pageUrl)
```

Comment:
Should be inside a `try..catch` as `pageUrl` can be undefined

Reply:
There's a guard a few lines up.

Reply:
After testing, it seems like there are a few values that could throw here. I'll add it.